multigeiger/display.cpp

```diff
@@ -0,0 +1,151 @@
+// OLED display related code
+
+#include <Arduino.h>
+#include <U8x8lib.h>
+
+#include "version.h"
+#include "log.h"
+#include "userdefines.h"
+
+#include "display.h"
+
+#if CPU == STICK
+#define PIN_DISPLAY_ON 25
+#endif
+
+#define PIN_OLED_RST 16
+#define PIN_OLED_SCL 15
+#define PIN_OLED_SDA 4
+
+#if CPU != STICK
+U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(PIN_OLED_RST, PIN_OLED_SCL, PIN_OLED_SDA);
+#else
+U8X8_SSD1306_64X32_NONAME_HW_I2C u8x8(PIN_OLED_RST, PIN_OLED_SCL, PIN_OLED_SDA);
+#endif
+
+bool displayIsClear;
+
+void display_start_screen(void) {
+  char line[20];
+
+  u8x8.clear();
+  #if CPU == STICK
+  // Display is only 4 lines by 8 characters; lines counting from 2 to 5
+  u8x8.setFont(u8x8_font_5x8_f);        // use really small font
+  for (int i = 2; i < 6; i++) {
+    u8x8.drawString(0, i, "        ");  // clear all 4 lines
+  }
+  #endif
+
+  #if CPU == STICK
+  u8x8.drawString(0, 2, "Geiger-");
+  u8x8.drawString(0, 3, " Counter");
+  u8x8.drawString(0, 4, "Version:");
+  snprintf(line, 9, "%s", VERSION_STR);  // 8 chars + \0 termination
+  u8x8.drawString(0, 5, line);
+  #else
+  u8x8.setFont(u8x8_font_7x14_1x2_f);
+  u8x8.println("Geiger-Counter");
+  u8x8.println("==============");
+  snprintf(line, 15, "%s", VERSION_STR);  // 14 chars + \0 termination
+  u8x8.setCursor(7 - strlen(line) / 2, 4);
+  u8x8.print(line);
+  u8x8.setCursor(1, 6);
+  u8x8.print("Info:boehri.de");
+  #endif
+
+  displayIsClear = false;
+};
+
+void setup_display(void) {
+  u8x8.begin();
+
+  #if CPU == STICK
+  pinMode(PIN_DISPLAY_ON, OUTPUT);
+  digitalWrite(PIN_DISPLAY_ON, HIGH);
+  #endif
+
+  display_start_screen();
+}
+
+void clearDisplayLine(int line) {
+  #if CPU != STICK
+  const char *blank = "                ";
+  #else
+  const char *blank = "        ";
+  #endif
+  u8x8.drawString(0, line, blank);
+}
+
+void displayStatusLine(String txt) {
+  #if CPU != STICK
+  int line = 7;
+  #else
+  int line = 5;
+  #endif
+  u8x8.setFont(u8x8_font_5x8_f);
+  clearDisplayLine(line);
+  u8x8.drawString(0, line, txt.c_str());
+}
+
+char *nullFill(int n, int digits) {
+  static char erg[9];  // max. 8 digits possible!
+  if (digits > 8) {
+    digits = 8;
+  }
+  char format[5];
+  sprintf(format, "%%%dd", digits);
+  sprintf(erg, format, n);
+  return erg;
+}
+
+void DisplayGMC(int TimeSec, int RadNSvph, int CPS, bool use_display, bool connected) {
+  if (!use_display && !displayIsClear) {
+    u8x8.clear();
+    clearDisplayLine(4);
+    clearDisplayLine(5);
```

Comment:
@rexfue do you know why the lower 2 lines are needed?

Reply:
I guess, it was because lines 4 and 5 where not cleared by u8x8.clear() at small display.

Reply:
but that would be an upstream bug in the library.

for such stuff a bug should be filed against the library code.

Reply:
https://github.com/ecocurious2/MultiGeiger/issues/208